napalm_eos/eos.py

```diff
@@ -1201,3 +1201,162 @@ def traceroute(self, destination, source='', ttl=0, timeout=0):
                 previous_probe_ip_address = ip_address
 
         return traceroute_result
+
+    def get_bgp_neighbors_detail(self, neighbor_address=''):
+        """Function to return detailed BGP information.
+
+        Information returned by this function can be about a single peer or
+        about all peers. This can be controlled using the following:
+
+        :params neighbor_address: Retuns the statistics for a specific
+                                  BGP neighbor or for all BGP neighbors.
+        """
+
+        peer_ver = None
+        is_all = True if not neighbor_address else False
+
+        if is_all:
+            command = ['show ip bgp neighbors vrf all']
+        else:
+            peer_ver = IPAddress(neighbor_address).version
+            if peer_ver == 4:
+                command = ['show ip bgp neighbors %s vrf all' %
+                           neighbor_address]
+            else:
+                command = ['show ipv6 bgp neighbors %s vrf all' %
+                           neighbor_address]
+
+        raw_output = (
+            self.device.run_commands(command, encoding='text')[0]['output'])
+
+        bgp_detail_info = {}
+        int_fields = ['accepted_prefix_count', 'local_as', 'remote_as',
+                      'local_port', 'remote_port', 'local_port',
+                      'input_messages', 'output_messages', 'input_updates',
+                      'output_updates', 'messages_queued_out', 'holdtime',
+                      'configured_holdtime', 'keepalive',
+                      'configured_keepalive', 'advertise_prefix_count',
+                      'received_prefix_count']
+
+        # Parsing through the raw bgp output
+        if peer_ver == 6:
+            neighbor_match_string = (
+                r'BGP neighbor is (.*), remote AS (\d+),')
+        else:
+            neighbor_match_string = (
+                r'BGP neighbor is (\d+.\d+.\d+.\d+), remote AS (\d+),')
+
+        parsed_peers = re.findall(neighbor_match_string, raw_output)
+
+        # Adding remote ASN's as top level keys in dict
+        for item in parsed_peers:
+            if item[1] not in bgp_detail_info:
+                bgp_detail_info[int(item[1])] = []
+
+        if is_all:
+            # Issuing commands per peer
+            for item in parsed_peers:
+                if peer_ver == 6:
+                    command = ['show ipv6 bgp neighbors %s vrf all' % item[0]]
```

Comment:
I see a problem with what you are suggesting. The output of `show ip bgp neighbors vrf all` does not have a conversion with eos so as to return the output in json. This makes parsing through 1000's of neighbors (if thats the case) very tedious. Hence the need of the per peer command in the event that the user using this method, does not specify the neighbor_address.

Secondly, the `show ip bgp neighbors vrf all` command does not return ipv6 neighbors and details about them. Hence, the need for checking the neighbor_address version and issuing the appropriate command.

I do understand that there are multiple requests being sent to the device so as to retrieve all the data, however, in my opinion, its better to have a slower function that returns all the data, rather than a faster one not returning the required data.

Please let me know your thoughts on the same. I am not refactoring this portion of the code for now.